modules/search-provision/src/main/scala/io/renku/search/provision/SolrRemovalProcess.scala

```diff
@@ -0,0 +1,154 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.renku.search.provision
+
+import cats.effect.{Async, Resource, Temporal}
+import cats.syntax.all.*
+import cats.{MonadThrow, Show}
+import fs2.Chunk
+import fs2.io.net.Network
+import io.github.arainko.ducktape.*
+import io.renku.avro.codec.{AvroDecoder, AvroReader}
+import io.renku.queue.client.{DataContentType, QueueClient, QueueMessage}
+import io.renku.redis.client.{ClientId, QueueName, RedisConfig}
+import io.renku.search.solr.client.SearchSolrClient
+import io.renku.solr.client.SolrConfig
+import org.apache.avro.Schema
+import scribe.Scribe
+import io.renku.search.solr.documents.DocumentId
+import scala.concurrent.duration.*
+import cats.data.NonEmptyList
+
+trait SolrRemovalProcess[F[_]]:
+  def removalProcess: F[Unit]
+
+object SolrRemovalProcess:
+  private val clientId: ClientId = ClientId("search-provisioner")
+
+  def make[F[_]: Async: Network: Scribe, In](
+      queueName: QueueName,
+      inSchema: Schema,
+      redisConfig: RedisConfig,
+      solrConfig: SolrConfig
+  )(using
+      Show[In],
+      Transformer[In, DocumentId],
+      AvroDecoder[In]
+  ): Resource[F, SolrRemovalProcess[F]] =
+    SearchSolrClient.make[F](solrConfig).map {
+      new SolrRemovalProcessImpl[F, In](
+        queueName,
+        inSchema,
+        clientId,
+        QueueClient.make[F](redisConfig),
+        _
+      )
+    }
+
+private class SolrRemovalProcessImpl[F[_]: Async: Scribe, In](
+    queueName: QueueName,
+    inSchema: Schema,
+    clientId: ClientId,
+    queueClientResource: Resource[F, QueueClient[F]],
+    solrClient: SearchSolrClient[F]
+)(using Show[In], Transformer[In, DocumentId], AvroDecoder[In])
+    extends SolrRemovalProcess[F]:
+  override def removalProcess: F[Unit] =
+    queueClientResource
+      .use { queueClient =>
+        findLastProcessed(queueClient) >>= { maybeLastProcessed =>
+          queueClient
+            .acquireEventsStream(queueName, chunkSize = 1, maybeLastProcessed)
+            .evalMap(decodeMessage(queueClient))
+            .evalTap(logInfo)
+            .groupWithin(chunkSize = 10, timeout = 500 millis)
+            .evalMap(deleteFromSolr(queueClient))
+            .compile
+            .drain
+            .handleErrorWith(logAndRestart)
+        }
+      }
+      .handleErrorWith(logAndRestart)
+
+  private def findLastProcessed(queueClient: QueueClient[F]) =
+    queueClient.findLastProcessed(clientId, queueName)
+
+  private lazy val logInfo: ((QueueMessage, Seq[In])) => F[Unit] = { case (m, v) =>
+    Scribe[F].info(
+      "Received mesage " + s"queue: $queueName, " +
+        s"id: ${m.id}, " +
+        s"source: ${m.header.source}, " +
+        s"type: ${m.header.`type`} " +
+        s"for: ${v.mkString_(", ")}"
+    )
+  }
+
+  private val avro = AvroReader(inSchema)
+
+  private def decodeMessage(queueClient: QueueClient[F])(
```

Comment:
I guess we could use the `QueueMessageDecoder` I introduced with the last PR to prevent code repetition. It's already used in both the insert and update routes.